src/smtencoding/FStar.SMTEncoding.Encode.fs

```diff
@@ -1413,9 +1471,15 @@ let encode_free_var env fv tt t_norm quals =
          else let encode_non_total_function_typ = lid.nsstr <> "Prims" in
               let formals, (pre_opt, res_t) =
                 let args, comp = curried_arrow_formals_comp t_norm in
+                let comp =
+                  if is_reifiable_comp env.tcenv comp
+                  then S.mk_Total (TcUtil.reify_comp ({env.tcenv with lax=true}) (U.lcomp_of_comp comp) U_unknown)
```

Comment:
I'm not sure this is the right thing to do in this case. Heads up @aseemr.

Consider something like:

```
assume val f : int -> St int

let g : St (int -> St int) = 
   let _ = alloc 0 in 
   f
```

@aseemr added encodings for the types of non-total functions specifically to capture such cases. 

But, this change seems to always encode the type of reifiable computations as its reified representation. That could work, but I'm not sure. Does this change work well with code such as the even counters example?